Adjustable asynchronous insert timeouts

jkartseva opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement auto-adjustment for asynchronous insert timeouts. The following settings are introduced: async_insert_poll_timeout_ms, async_insert_use_adaptive_busy_timeout, async_insert_busy_timeout_min_ms, async_insert_busy_timeout_max_ms, async_insert_busy_timeout_increase_rate, async_insert_busy_timeout_decrease_rate.

async_insert_busy_timeout_ms is aliased to async_insert_busy_timeout_max_ms.
### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

jkartseva committed:
Adaptive asynchronous insert timeouts

Implement the algorithm described in #56783 for adaptive asynchronous
insert timeouts.
- The adaptive async insert timeout can take values within
[async_insert_busy_timeout_min_ms, async_insert_busy_timeout_max_ms].
- The initial value is set to async_insert_busy_timeout_min_ms.
- If the elapsed time since the most recent queue insert was
  greater than the maximum timeout, process the queue content immediately,
  and reduce the timeout.
- If the elapsed time was long enough (longer than a would-be decreased
  timeout), decrease the timeout.
- The adaptive timeout is changes exponentially based on the
  async_insert_busy_timeout_{increase|decrease}_rate.

Fixes: https://github.com/ClickHouse/ClickHouse/issues/56783

jkartseva committed:
Asynchronous insert queue size and bytes metrics

Add AsynchronousInsertQueueSize and AsynchronousInsertQueueBytes
metrics to improve observability of asynchronous inserts.
The metrics do not account for tasks dispatched for immediate processing
(as opposed to, e.g., PendingAsyncInsert).

```
SELECT value
FROM system.metrics
WHERE metric IN ('AsynchronousInsertQueueSize', 'PendingAsyncInsert')

Query id: a711dd83-b48d-4ad5-8031-fa59b21a7c38

┌─value─┐
│    18 │
│    23 │
└───────┘
```

```
SELECT value
FROM system.metrics
WHERE metric IN ('AsynchronousInsertQueueSize', 'AsynchronousInsertQueueBytes')

Query id: b35a7ceb-2bb5-46ad-b301-e6cf03508699

┌─value─┐
│    28 │
│  1372 │
└───────┘
```

jkartseva committed:
Integration tests for adaptive async insert timeout

Add initial tests.

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #58486 from jkartseva/ClickHouse:adjustable-async-timeouts -->
*This is an automated comment for commit 4309bba5ceb481add10dbe465b8c500defc0f9f7 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58486/4309bba5ceb481add10dbe465b8c500defc0f9f7/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58486/4309bba5ceb481add10dbe465b8c500defc0f9f7/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

qoega commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -323,19 +338,24 @@ AsynchronousInsertQueue::pushDataChunk(ASTPtr query, DataChunk chunk, ContextPtr
-            auto timeout = now + Milliseconds{key.settings.async_insert_busy_timeout_ms};
-            it->second = shard.queue.emplace(timeout, Container{key, std::make_unique<InsertData>()}).first;
+            if (getContext()->getSettingsRef().allow_experimental_async_insert_adjustable_busy_timeout)
+                updateBusyWaitTimeout(now, shard, elapsed_time_exceeds_max_timeout, shard_num);
```
I do not get if it, how it works. We update timeout before we check bytes/queries limit. 
Will we update busy timeout in this case as well? As I understand the algorithm we do not need to batch more if there is enough bytes per batch.

jkartseva replied:
We do not mark the whole queue shard for immediate processing, but rather only the entries with the specific `key.hash` that have reached either bytes or # of queries limit:

https://github.com/ClickHouse/ClickHouse/blob/06274c7bc641b10038a696549e384b756654ccea/src/Interpreters/AsynchronousInsertQueue.cpp#L326-L354

Other entries that have fallen into the same queue shard will remain in the async queue. 
Hence, I believe, we should adjust the timeout.

jkartseva commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -359,10 +380,23 @@ AsynchronousInsertQueue::pushDataChunk(ASTPtr query, DataChunk chunk, ContextPtr
+        {
+            if (inserted)
+                CurrentMetrics::add(CurrentMetrics::AsynchronousInsertQueueSize);
+            CurrentMetrics::add(CurrentMetrics::AsynchronousInsertQueueBytes, entry_data_size);
```
`size_in_bytes` corresponds to the entries with the same `key.hash`, they have already been inserted and accounted for by the metric. Here, we should only increase the metric by the size of the current entry (`entry_data_size`). 
However, we should subtract the size of the previously inserted entries from the metric if the entreis are to be processed synchronously.

jkartseva commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -181,6 +183,20 @@ AsynchronousInsertQueue::AsynchronousInsertQueue(ContextPtr context_, size_t poo
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS, "'async_insert_busy_timeout_ms' must be greater than 'async_insert_busy_timeout_min_ms'");
+
+    if (Milliseconds(settings.async_insert_busy_timeout_min_ms) == std::chrono::milliseconds::zero())
```
This can be done; I would still like to preserve milliseconds as the smallest time unit, without dispersing into microseconds.

jkartseva committed:
Log timeout for entries processed asynchronously

jkartseva committed:
Adaptive async timeouts: consider queue flush history

In addition to the time since the most recent insert,
consider the elapsed time between the two recent queue
flushes when decreasing the timeout or processing an
entry synchronously.

jkartseva committed:
Introduce polling timeout for asynchronous queue

The current timeout for checking updates in the asynchronous queue is
equal to the timeout used for queue entry
(async_insert_busy_timeout_ms).
That means that, in the worst case, an entry spends twice the time of the
asynchronous timeout in the queue.

jkartseva commented on src/Core/Settings.h:
```diff
@@ -746,7 +746,13 @@ class IColumn;
-    M(Milliseconds, async_insert_busy_timeout_ms, 200, "Maximum time to wait before dumping collected data per query since the first data appeared", 0) \
+    M(Milliseconds, async_insert_busy_timeout_ms, 200, "Maximum time to wait before dumping collected data per query since the first data appeared. If auto-adjusting is enabled through enable_async_insert_adaptive_busy_timeout, the setting corresponds to the initial time to wait, and the following wait time will be auto-adjusted", 0) \
+    M(Milliseconds, async_insert_poll_timeout_ms, 50, "Timeout for polling data from asynchronous insert queue", 0) \
+    M(Bool, enable_async_insert_adaptive_busy_timeout, false, "Enable adaptive busy timeouts for asynchronous inserts", 0) \
```
I wanted this setting to be deployed and adopted gradually, so I added `enable_async_insert_adaptive_busy_timeout` with a kill switch in mind. We might meet a corner case not accounted for during testing.
Would `allow_experimental...` be a better name for the purpose? The setting can be marked as deprecated later on.

antonio2368 replied:
Hm I see this more as an implementation detail for async inserts and a strictly better implementation.
From users perspective, he should worry only if he wants to use async inserts or not. I'm okay with having this flag as a safety net if the new implementation misbehaves but I cannot see a reason why someone would want the old behavior.

jkartseva replied:
Ready for review. Tests failures are not relevant, e.g., https://s3.amazonaws.com/clickhouse-test-reports/0/0cbadc9b52f5e728a397b3081c6887c0f4213856/stress_test__debug_.html is also failing on master.
I'll take a look at what might be wrong there.

Currently, the new feature is disabled by default, we may choose to enable it. I want to preserve `allow_experimental_<..>` kill switch, though.

jkartseva commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -217,14 +230,14 @@ AsynchronousInsertQueue::~AsynchronousInsertQueue()
-    });
+    pool.scheduleOrThrowOnError(
+        [this, key, global_context, shard_num, my_data = std::make_shared<InsertDataPtr>(std::move(data))]() mutable
+        { processData(key, std::move(*my_data), std::move(global_context), std::ref(flush_time_history_per_queue_shard[shard_num])); });
```
an artifact, will remove

jkartseva commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -429,17 +567,21 @@ void AsynchronousInsertQueue::processBatchDeadlines(size_t shard_num)
+                = std::min(shard.busy_timeout_ms, Milliseconds(getContext()->getSettingsRef().async_insert_poll_timeout_ms));
+            shard.are_tasks_available.wait_for(
+                lock,
+                rel_time,
```
That's how it's implemented right now. Imagine an entry arriving in the queue just after a timeout has fired. When the timeout fires again, that entry won't be processed. Instead, it will stay in the queue until the next cycle. Essentially, that entry will remain in the queue for almost double the intended time.
If entries arrive uniformly over time, they are expected to remain in the queue for an additional half of the timeout.
This is especially visible for large (over 5s) timeouts and frequent inserts.

jkartseva replied:
... e.g., the following tests:
```
find tests/ -name '*.sh' -o -name '*.sql' 2>/dev/null | xargs grep 'async_insert_busy_timeout.*00000'
tests/queries/0_stateless/02015_async_inserts_2.sh:url="${CLICKHOUSE_URL}&async_insert=1&wait_for_async_insert=1&async_insert_busy_timeout_ms=600000&async_insert_max_query_number=3&async_insert_deduplicate=1"
tests/queries/0_stateless/02726_async_insert_flush_queue.sql:SET async_insert_busy_timeout_ms = 1000000;
tests/queries/0_stateless/02884_async_insert_native_protocol_1.sh:async_insert_options="--async_insert 1 --wait_for_async_insert 0 --async_insert_busy_timeout_ms 1000000"
tests/queries/0_stateless/02884_async_insert_native_protocol_3.sh:async_insert_options="--async_insert 1 --wait_for_async_insert 0 --async_insert_busy_timeout_ms 1000000"
tests/queries/0_stateless/02884_async_insert_skip_settings.sql:SET async_insert_busy_timeout_ms = 100000;
```
The reason why they don't fail is that the condvar polling time is taken from the server settings (default 200ms), not the query settings.

antonio2368 replied:
okay, that makes sense, thanks for explaining

jkartseva replied:
Instead of introducing an additional `async_insert_poll_timeout_ms` setting, we may use a fixed fraction (e.g., 1/8) of `busy_timeout_ms`.

antonio2368 replied:
no, I think it will make it worse because tuning it will be abstract
I think polling makes sense and we can leave the setting so you can always fine tune it

jkartseva commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -176,11 +179,21 @@ AsynchronousInsertQueue::AsynchronousInsertQueue(ContextPtr context_, size_t poo
+    const auto & settings = getContext()->getSettingsRef();
+
+    for (size_t i = 0; i < pool_size; ++i)
+        queue_shards[i].busy_timeout_ms = settings.async_insert_busy_timeout_ms;
```
I've been considering a smooth transition from the state 'enable_async_insert_adaptive_busy_timeout=0' to '1'. async_insert_busy_timeout_ms is a timeout anticipated by a customer, and the adaptive algorithm treats it as the initial point. It minimizes the value if inserts are not frequent.

antonio2368 replied:
> is a timeout anticipated by a customer

if they touch for some reason the timeout setting, otherwise they don't anticipate anything as long as it works

the problem here is that we would like to solve the timeout of the initial insert, and if the user set `async_insert_busy_timeout_ms` to some high value, it will take a long time for even a simple insert of 1 row.
IMO we should start from minimal flushtimeout and gradually grow from there if the insert rate increases.

antonio2368 commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -374,6 +414,103 @@ AsynchronousInsertQueue::pushDataChunk(ASTPtr query, DataChunk chunk, ContextPtr
+    const auto min_ms = std::chrono::milliseconds(settings.async_insert_busy_timeout_min_ms);
+    const auto max_ms = std::chrono::milliseconds(settings.async_insert_busy_timeout_max_ms);
+
+    if (t_ms < min_ms)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting 'async_insert_busy_timeout_ms' must be greater or equal than 'async_insert_busy_timeout_min_ms'");
+
+    if (t_ms > max_ms)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting 'async_insert_busy_timeout_ms' must be less or equal than 'async_insert_busy_timeout_max_ms'");
```
ah I can't read both code and english...

antonio2368 commented on src/Core/Settings.h:
```diff
@@ -746,7 +746,13 @@ class IColumn;
+    M(Milliseconds, async_insert_poll_timeout_ms, 50, "Timeout for polling data from asynchronous insert queue", 0) \
+    M(Bool, enable_async_insert_adaptive_busy_timeout, false, "Enable adaptive busy timeouts for asynchronous inserts", 0) \
+    M(Milliseconds, async_insert_busy_timeout_min_ms, 50, "Minimum value of an asynchronous insert timeout if auto-adjusting is enabled through enable_async_insert_adaptive_busy_timeout", 0) \
+    M(Milliseconds, async_insert_busy_timeout_max_ms, 1000, "Maximum value of an asynchronous insert timeout if auto-adjusting is enabled through enable_async_insert_adaptive_busy_timeout", 0) \
```
shouldn't this be same as `async_insert_busy_timeout_ms`?
in the current implementation `async_insert_busy_timeout_ms` is the total time limit which will force a flush, i.e. maximum we can keep the entries before flushing.
Also, what if user changed `async_insert_busy_timeout_ms` to 2s? He would get errors how `async_insert_busy_timeout_max_ms` is not okay, and then he needs to figure out what it is and what value to set it to.

Why not put `async_insert_busy_timeout_ms` as alias to `async_insert_busy_timeout_max_ms` to keep backward compatibility?

jkartseva replied:
This sounds good to me. Is there a way to alias two settings (async_insert_busy_timeout_ms and async_insert_busy_timeout_max_ms) so that they are tied to the same value?

jkartseva replied:
TIL `ALIAS` expression, will do.

jkartseva commented:
Tthe newly added metrics

<img width="1461" alt="Screenshot 2024-01-25 at 6 48 21 PM" src="https://github.com/ClickHouse/ClickHouse/assets/25540979/ab5f6b5c-9fe9-40a7-9f9c-f7be2b595030">


``` sql
SELECT 
  toStartOfInterval(event_time, INTERVAL {rounding:UInt32} SECOND)::INT AS t,
  avg(metric)
FROM (
  SELECT event_time, sum(CurrentMetric_AsynchronousInsertQueueBytes) AS metric 
  FROM clusterAllReplicas(default, system.metric_log)
  WHERE event_date >= toDate(now() - {seconds:UInt32})
    AND event_time >= now() - {seconds:UInt32}
  GROUP BY event_time)
GROUP BY t
ORDER BY t WITH FILL STEP {rounding:UInt32} SETTINGS skip_unavailable_shards = 1
```

``` sql
SELECT 
  toStartOfInterval(event_time, INTERVAL {rounding:UInt32} SECOND)::INT AS t,
  avg(value)
FROM (
  SELECT event_time, avg(timeout_milliseconds) AS value 
  FROM clusterAllReplicas(default, system.asynchronous_insert_log)
  WHERE event_date >= toDate(now() - {seconds:UInt32})
    AND event_time >= now() - {seconds:UInt32}
  GROUP BY event_time)
GROUP BY t
ORDER BY t WITH FILL STEP {rounding:UInt32} SETTINGS skip_unavailable_shards = 1
```

jkartseva committed:
Update settings.md

antonio2368 reviewed (commented):
Everything LGTM, it works really well.
Let's enable it by default and see if it breaks any other tests with async insert.

We still need to figure out what we want to do with the setting in general, I'm still in favor of keeping it enabled and rename `allow_experimental_async_insert_adaptive_busy_timeout` to something like `async_insert_use_adaptive_busy_timeout`.

jkartseva committed:
Update existing tests

antonio2368 commented on tests/queries/0_stateless/02134_async_inserts_formats.sh:
```diff
@@ -4,7 +4,7 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 . "$CURDIR"/../shell_config.sh
 
-url="${CLICKHOUSE_URL}&async_insert=1&wait_for_async_insert=1"
+url="${CLICKHOUSE_URL}&async_insert=1&wait_for_async_insert=1&async_insert_use_adaptive_busy_timeout=0"
```
why do you have to disable it here?

jkartseva replied:
The test became flaky after enabling adaptive timeouts by default:
https://pastila.nl/?01884edd/3302a987b576ef7e75fbc3d8e1e0bf3d#Dgf6yOJGn0Bk35nJDt0++g==

I suspect this is because a queue can be flushed synchronously, depending on how much time has elapsed since the last insert.

antonio2368 replied:
> I suspect this is because a queue can be flushed synchronously

yes, that makes sense

jkartseva replied:
Thanks for the review. I'll wait for the current workflow to end, then modify the test.

jkartseva replied:
Same for `async_insert_dedup_replicated_collapsing.sh` test: duplicate entries are flushed in separate batches if the timeout is small. So I disabled the feature for that test.

antonio2368 commented:
We can maybe add `async_insert_use_adaptive_busy_timeout` to random settings in `clickhouse-test`?

jkartseva commented:
I'm confused about "New settings are not reflected in settings changes history (see new_settings.txt)", though I have updated `settings.md`

https://s3.amazonaws.com/clickhouse-test-reports/58486/da12bf77cfeceef9476c29c447ef82acacf6cb8d/upgrade_check__debug_.html

Do you know by any chance where this `new_settings.txt` is located? @antonio2368

antonio2368 commented:
@jkartseva that file is generated by a query during upgrade test run to tell you which new settings were not added to `src/Core/SettingsChangesHistory.h`
Go to the failed test and open `new_settings.txt`
https://s3.amazonaws.com/clickhouse-test-reports/58486/da12bf77cfeceef9476c29c447ef82acacf6cb8d/upgrade_check__debug_.html

jkartseva committed:
Update SettingsChangesHistory.h

jkartseva commented:
> We can maybe add `async_insert_use_adaptive_busy_timeout` to random settings in `clickhouse-test`?

I'll publish a follow-up PR.

jkartseva commented:
Set version to 24.2 in settings history. It should be good to merge.

antonio2368 commented on src/Interpreters/AsynchronousInsertQueue.cpp:
```diff
@@ -374,6 +427,94 @@ AsynchronousInsertQueue::pushDataChunk(ASTPtr query, DataChunk chunk, ContextPtr
+    /// Adaptive timeout settings.
+    const auto min_ms = std::chrono::milliseconds(settings.async_insert_busy_timeout_min_ms);
+
+    if (min_ms > t_ms)
```
what will we do then for backward compatibility if someone set `async_insert_busy_timeout_ms` to a value less then default `async_insert_busy_timeout_min_ms`?

jkartseva replied:
Yes, it slipped my mind to update it, thanks. In that case, the minimum value of the two is taken as the async timeout.

jkartseva commented:
The adaptive timeout ramping up from the minimum could contradict the `async_insert_deduplicate` setting, whose intention is to eliminate duplicates in a batch. At the very least, synchronous queue flush should be disabled for the setting. I'll file a follow-up.

antonio2368 commented:
> The adaptive timeout ramping up from the minimum could contradict the `async_insert_deduplicate` setting, whose intention is to eliminate duplicates in a batch. At the very least, synchronous queue flush should be disabled for the setting. I'll file a follow-up.

No, deduplication eliminates duplicates in general (or at least in `replicated_deduplication_window_for_async_inserts` window).
Removing duplicates from batch is just optimization so we don't have to do it after verifying with Keeper.

qoega commented on tests/integration/test_async_insert_adaptive_busy_timeout/test.py:
```diff
@@ -0,0 +1,372 @@
+            fixed_tm_table_name,
+            fixed_tm_settings,
+            thread_num=15,
+            tasks=1000,
```
@jkartseva can you help me understand why you check with fixed timeout 1000 iterations, but compare it with 100 of adaptive timeout?
